File: 19. Remove Nth Node From End of List Medium/solution.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let first: ListNode | null = head;
    let second: ListNode | null = head;

    for (let i = 0; i < n; i++) {
        if (first === null) {
            return head;
        }
        first = first.next;
    }

    if (first === null) {
        return head !== null ? head.next : null;
    }

    while (first.next !== null) {
        first = first.next;
        second = second!.next;
    }

    if (second !== null && second.next !== null) {
        second.next = second.next.next;
    }

    return head;
}

File: 19. Remove Nth Node From End of List Medium/solution.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Val: 0}
	dummy.Next = head
	first := dummy
	second := dummy

	for i := 0; i < n+1; i++ {
		first = first.Next
	}

	for first != nil {
		first = first.Next
		second = second.Next
	}

	second.Next = second.Next.Next

	return dummy.Next
}
